pubtools/pulplib/_impl/fake/client.py

```diff
@@ -79,6 +82,52 @@ def get_repository(self, repository_id):
 
         return f_return(data[0])
 
+    def _do_upload_file(self, upload_id, file_obj, name):
+        # pylint: disable=unused-argument
+        is_file_obj = "close" in dir(file_obj)
+        if not is_file_obj:
+            file_obj = open(file_obj, "rb")
+
+        def do_next_upload(checksum, size):
+            data = file_obj.read(1024 * 1024)
+            if data:
+                checksum.update(data)
+                size += len(data)
+                return do_next_upload(checksum, size)
+            return f_return((checksum.hexdigest(), size))
+
+        out = f_flat_map(f_return(), lambda _: do_next_upload(hashlib.sha256(), 0))
+
+        if not is_file_obj:
+            out.add_done_callback(lambda _: file_obj.close())
+
+        return out
+
+    def _request_upload(self):
+        upload_request = {
+            "_href": "/pulp/api/v2/content/uploads/%s/" % self._request_id(),
+            "upload_id": "%s" % self._request_id,
```

Comment:
This doesn't look right, as you didn't call self._request_id here, so the "%s" would be formatted as e.g. 

```
<function ... at 0x7fa2a7...>
```

Which leads me to question: why is everything apparently working even though you've got a junk upload_id here?

I guess it's because upload_id was ignored in _do_upload_file.  In terms of the public API, seems like this doesn't matter at all, but to protect against this kind of bug you might want to do some check of upload_id in that method.